### PR TITLE
[feature](Nereids) Support OneRowRelation and EmptyRelation

### PR DESCRIPTION
**Pull request by 924060929**
# Proposed changes

Support OneRowRelation and EmptyRelation.

- OneRowRelation: `select 100, 'abc', substring('abc', 1, 2)`
- EmptyRelation: `select * from tbl limit 0`

Note:
PhysicalOneRowRelation will translate to UnionNode(constExpr) for BE execution

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [x] No
    - [ ] I don't know
2. Has unit tests been added:
    - [x] Yes
    - [ ] No
    - [ ] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [x] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [x] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [x] No

### CODE REVIEW

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/BindFunction.java [hunk 1]
```diff
@@ -52,6 +53,22 @@ public class BindFunction implements AnalysisRuleFactory {
     @Override
     public List<Rule> buildRules() {
         return ImmutableList.of(
+            RuleType.BINDING_ONE_ROW_RELATION_FUNCTION.build(
+                logicalOneRowRelation().then(project -> {
```
Comment (wangshuo128) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/BindFunction.java:57]:
```suggestion
                logicalOneRowRelation().then(oneRowRelation -> {
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/analyzer/UnboundOneRowRelation.java [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+
+/**
+ * A relation that contains only one row consist of some constant expressions.
+ * e.g. select 100, 'value'
+ */
+public class UnboundOneRowRelation extends LogicalLeaf implements Unbound, OneRowRelation {
+    private List<NamedExpression> projects;
```
Comment (wangshuo128) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/analyzer/UnboundOneRowRelation.java:46]:
```suggestion
    private final List<NamedExpression> projects;
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/Substring.java [hunk 3]
```diff
@@ -54,7 +54,7 @@ public Substring(Expression str, Expression pos) {
 
     @Override
     public DataType getDataType() {
-        return StringType.INSTANCE;
+        return VarcharType.SYSTEM_DEFAULT;
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/Substring.java:57]: could `new VarcharType(len)` works?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/analyzer/UnboundOneRowRelation.java [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    private final List<NamedExpression> projects;
+
+    public UnboundOneRowRelation(List<NamedExpression> projects) {
+        this(projects, Optional.empty(), Optional.empty());
+    }
+
+    private UnboundOneRowRelation(List<NamedExpression> projects, Optional<GroupExpression> groupExpression,
+            Optional<LogicalProperties> logicalProperties) {
+        super(PlanType.LOGICAL_UNBOUND_ONE_ROW_RELATION, groupExpression, logicalProperties);
+        Preconditions.checkArgument(projects.stream().allMatch(Expression::isConstant),
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/analyzer/UnboundOneRowRelation.java:55]: i doubt `isConstant` is too strictness. for example
```
SELECT now(), random();
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/analyzer/UnboundOneRowRelation.java [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+    }
+
+    @Override
+    public List<NamedExpression> getProjects() {
+        return projects;
+    }
+
+    @Override
+    public List<Expression> getExpressions() {
+        throw new UnsupportedOperationException(this.getClass().getSimpleName() + " don't support getExpression()");
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/analyzer/UnboundOneRowRelation.java:72]: why could not return projects?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/glue/translator/PhysicalPlanTranslator.java [hunk 6]
```diff
@@ -227,6 +232,56 @@ public PlanFragment visitPhysicalAggregate(
         return currentFragment;
     }
 
+    @Override
+    public PlanFragment visitPhysicalEmptyRelation(PhysicalEmptyRelation emptyRelation, PlanTranslatorContext context) {
+        List<Slot> output = emptyRelation.getOutput();
+        TupleDescriptor tupleDescriptor = generateTupleDesc(output, null, context);
+        for (int i = 0; i < output.size(); i++) {
+            SlotDescriptor slotDescriptor = tupleDescriptor.getSlots().get(i);
+            slotDescriptor.setIsNullable(true);
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/glue/translator/PhysicalPlanTranslator.java:241]: need same comment as OneRowRelation

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/Substring.java (hunk 3 above)
Comment (924060929) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/Substring.java:57]: keep the same type as OriginPlanner, or else BE can not found the function

File: fe/fe-core/src/main/java/org/apache/doris/nereids/analyzer/UnboundOneRowRelation.java (hunk 4 above)
Comment (924060929) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/analyzer/UnboundOneRowRelation.java:55]: I see, thinks

File: fe/fe-core/src/main/java/org/apache/doris/nereids/analyzer/UnboundOneRowRelation.java (hunk 5 above)
Comment (924060929) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/analyzer/UnboundOneRowRelation.java:72]: Keep same as the logic: UnboundXxx can not invoke getExpressions(), like UnboundRelation.
It maybe help up to find some bug if we getExpression from an unbound plan.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/glue/translator/PhysicalPlanTranslator.java (hunk 6 above)
Comment (924060929) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/glue/translator/PhysicalPlanTranslator.java:241]: ok

File: fe/fe-core/src/main/java/org/apache/doris/nereids/analyzer/UnboundOneRowRelation.java (hunk 4 above)
Comment (924060929) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/analyzer/UnboundOneRowRelation.java:55]: 'not contains any slot' is correct?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/Substring.java (hunk 3 above)
Comment (morrySnow) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/Substring.java:57]: if be can obtain length, so optimize in runtime could be used. Such as low cardinality optimization in building hash table

File: fe/fe-core/src/main/java/org/apache/doris/nereids/analyzer/UnboundOneRowRelation.java (hunk 4 above)
Comment (morrySnow) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/analyzer/UnboundOneRowRelation.java:55]: > 'not contains any slot' is correct?

i think so

File: fe/fe-core/src/main/java/org/apache/doris/nereids/analyzer/UnboundOneRowRelation.java (hunk 5 above)
Comment (morrySnow) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/analyzer/UnboundOneRowRelation.java:72]: just curiousness, i think current code is ok

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/Substring.java (hunk 3 above)
Comment (924060929) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/Substring.java:57]: so you mean this?
```java
    @Override
    public DataType getDataType() {
        if (getLength() instanceof IntegerLiteral) {
            return VarcharType.createVarcharType(((IntegerLiteral) getLength()).getValue());
        }
        return VarcharType.SYSTEM_DEFAULT;
    }
```

**Review by morrySnow — APPROVED** [review]

**Pull request merged**